src/libraries/System.Resources.Extensions/src/BinaryFormat/BinaryFormattedObject.TypeResolver.cs

```diff
@@ -0,0 +1,139 @@
+        private readonly FormatterAssemblyStyle _assemblyMatching;
+        private readonly SerializationBinder? _binder;
+
+        private readonly Dictionary<string, Assembly> _assemblies = [];
+        private readonly Dictionary<string, Type> _types = [];
+
+        internal DefaultTypeResolver(Options options)
+        {
+            _assemblyMatching = options.AssemblyMatching;
+            _binder = options.Binder;
+        }
+
+        /// <summary>
+        ///  Resolves the given type name against the specified library.
+        /// </summary>
+        [RequiresUnreferencedCode("Calls System.Reflection.Assembly.GetType(String)")]
+        [return: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)]
+        Type ITypeResolver.GetType(TypeName typeName)
+        {
+            Debug.Assert(typeName.AssemblyName is not null);
+
+            if (_types.TryGetValue(typeName.AssemblyQualifiedName, out Type? cachedType))
+            {
+                return cachedType;
+            }
+
+            if (_binder?.BindToType(typeName.AssemblyName.FullName, typeName.FullName) is Type binderType)
+            {
+                // BinaryFormatter is inconsistent about what caching behavior you get with binders.
+                // It would always cache the last item from the binder, but wouldn't put the result
+                // in the type cache. This could lead to inconsistent results if the binder didn't
+                // always return the same result for a given set of strings. Choosing to always cache
+                // for performance.
+
+                _types[typeName.AssemblyQualifiedName] = binderType;
+                return binderType;
+            }
+
+            if (!_assemblies.TryGetValue(typeName.AssemblyName.FullName, out Assembly? assembly))
+            {
+                AssemblyName assemblyName = typeName.AssemblyName.ToAssemblyName();
+                try
+                {
+                    assembly = Assembly.Load(assemblyName);
+                }
+                catch
+                {
+                    if (_assemblyMatching != FormatterAssemblyStyle.Simple)
+                    {
+                        throw;
+                    }
+
+                    assembly = Assembly.Load(assemblyName.Name!);
+                }
+
+                _assemblies.Add(typeName.AssemblyName.FullName, assembly);
+            }
+
+            Type? type = _assemblyMatching != FormatterAssemblyStyle.Simple
+                ? assembly.GetType(typeName.FullName)
+                : GetSimplyNamedTypeFromAssembly(assembly, typeName);
+
+            _types[typeName.AssemblyQualifiedName] = type ?? throw new SerializationException($"Could not find type '{typeName}'.");
+            return type;
+        }
+
+        [RequiresUnreferencedCode("Calls System.Reflection.Assembly.GetType(String, Boolean, Boolean)")]
+        private static Type? GetSimplyNamedTypeFromAssembly(Assembly assembly, TypeName typeName)
```

Comment:
Note that this is effectively a straight copy of the existing `BinaryFormatter` logic.